### PR TITLE
Remove cudf._lib.groupby in favor of inlining pylibcudf

### PR DESCRIPTION
**Pull request by mroeschke**
## Description
Contributes to https://github.com/rapidsai/cudf/issues/17317

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/groupby/groupby.py [hunk 1]
```diff
@@ -582,11 +700,176 @@ def rank(x):
...
+                            any(
+                                a in str_agg
+                                for a in {"count", "max", "min", "unique"}
+                            )
+                        )
+                    )
+                ):
+                    raise TypeError(
+                        f"{col.dtype} type does not support {agg} operations"
+                    )
```
Comment (Matt711) [inline python/cudf/cudf/core/groupby/groupby.py:804]: Would you be okay with moving this logic out into separate functions that determine if an aggregation is not supported for string and categorical types? They can be nested in `_aggregate`. Eg.
```python
def _is_unsupported_for_string(col, str_agg):
    cumulative_agg = str_agg in {"cumsum", "cummin", "cummax"}
    basic_agg = any(a in str_agg for a in {
        "count", "max", "min", "first", "last", "nunique", "unique", "nth"
    })
    return (
        is_string_dtype(col)
        and str_agg not in _STRING_AGGS
        and (cumulative_agg or (not basic_agg and str_agg != "list"))
    )

def _is_unsupported_for_categorical(col, str_agg):
    cumulative_agg = atr_agg in {"cumsum", "cummin", "cummax"}
    basic_agg = any(a in str_agg for a in {"count", "max", "min", "unique"})
    return (
        _is_categorical_dtype(col)
        and agg_str not in _CATEGORICAL_AGGS
        and (cumulative_agg or not basic_agg)
    )
```

Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py:804]: Maybe use a `singledispatch` function on the dtype.

Comment (mroeschke) [reply python/cudf/cudf/core/groupby/groupby.py]: Sure thing. Was able adapt this to use `singledispatch`

File: python/cudf/cudf/core/groupby/groupby.py [hunk 2]
```diff
@@ -46,6 +57,152 @@
...
+        a in str_agg for a in ("count", "max", "min", "unique")
+    )
+    return str_agg not in _CATEGORICAL_AGGS and (
+        cumulative_agg or not_basic_agg
+    )
+
+
+def _is_all_scan_aggregate(all_aggs: list[list[str]]) -> bool:
+    """
+    Returns true if all are scan aggregations.
```
Comment (Matt711) [inline python/cudf/cudf/core/groupby/groupby.py:170]: nitpick
```suggestion
    Returns True if all are scan aggregations.
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 3]
```diff
@@ -582,11 +739,137 @@ def rank(x):
...
+                    [
+                        col.to_pylibcudf(mode="read")
+                        for col in self.grouping.keys._columns
+                    ]
+                ),
+                plc.types.NullPolicy.EXCLUDE
+                if self._dropna
+                else plc.types.NullPolicy.INCLUDE,
+            )
+            # Do we need this because we just check _spill_locks in test_spillable_df_groupby?
```
Comment (Matt711) [inline python/cudf/cudf/core/groupby/groupby.py:755]: Should we rephrase this as a note?
```suggestion
            # Note: We return a SimpleNamespace with an additional _spill_locks
            # attribute solely to verify that the spill_lock is set correctly.
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 4]
```diff
@@ -582,11 +739,137 @@ def rank(x):
...
+                    raise TypeError(
+                        f"{col.dtype} type does not support {agg} operations"
+                    )
+                agg_obj = aggregation.make_aggregation(agg)
+                if (
+                    valid_aggregations == "ALL"
+                    or agg_obj.kind in valid_aggregations
+                ):
+                    included_aggregations_i.append((agg, agg_obj.kind))
+                    col_aggregations.append(agg_obj.c_obj)
```
Comment (Matt711) [inline python/cudf/cudf/core/groupby/groupby.py:812]: Can we call this a `plc_obj`? And change this line too https://github.com/rapidsai/cudf/blob/branch-25.02/python/cudf/cudf/core/_internals/aggregation.py#L32
```suggestion
                    col_aggregations.append(agg_obj.plc_obj)
```

**Review by Matt711 — APPROVED** [review]
Thanks, minor suggestions. I also like the `singledispatch` function you added.

File: python/cudf/cudf/core/groupby/groupby.py (hunk 3 above)
Comment (mroeschke) [reply python/cudf/cudf/core/groupby/groupby.py:755]: Going to keep this one phrased as a question. I _think_ if this is only used for testing (and is maybe unnecessary) we could potentially just remove this in the future
